src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs

```diff
@@ -0,0 +1,337 @@
+            }
+            _ = packageSources ?? throw new ArgumentNullException(nameof(packageSources));
+
+
+            ConcurrentDictionary<PackageSource, IEnumerable<IPackageSearchMetadata>> searchResults = new ConcurrentDictionary<PackageSource, IEnumerable<IPackageSearchMetadata>>();
+            await Task.WhenAll(packageSources.Select(async source =>
+                {
+                    IEnumerable<IPackageSearchMetadata> foundPackages = await GetPackageMetadataAsync(source, packageIdentifier, includePrerelease: false, cancellationToken).ConfigureAwait(false);
+                    if (foundPackages == null)
+                    {
+                        return;
+                    }
+                    searchResults[source] = foundPackages;
+                })).ConfigureAwait(false);
+
+            if (!searchResults.Any())
+            {
+                throw new InvalidNuGetSourceException("Failed to load NuGet sources", packageSources.Select(source => source.Source));
+            }
+
+            var accumulativeSearchResults = searchResults
+                .SelectMany(result => result.Value.Select(p => new { source = result.Key, package = p }));
+
+            if (!accumulativeSearchResults.Any())
+            {
+                _nugetLogger.LogWarning($"{packageIdentifier} is not found in NuGet feeds {string.Join(", ", packageSources.Select(source => source.Source))}.");
+                throw new PackageNotFoundException(packageIdentifier, packageSources.Select(source => source.Source));
+            }
+
+            var latestVersion = accumulativeSearchResults.Aggregate(
+                (current, max) =>
+                {
+                    if (max == null) return current;
+                    return current.package.Identity.Version > max.package.Identity.Version ? current : max;
+                });
+            return (latestVersion.source, latestVersion.package);
+        }
+
+        private async Task<(PackageSource, IPackageSearchMetadata)> GetPackageMetadataAsync(string packageIdentifier, NuGetVersion packageVersion, IEnumerable<PackageSource> sources, CancellationToken cancellationToken)
+        {
+            if (string.IsNullOrWhiteSpace(packageIdentifier))
+            {
+                throw new ArgumentException($"{nameof(packageIdentifier)} cannot be null or empty", nameof(packageIdentifier));
+            }
+            _ = packageVersion ?? throw new ArgumentNullException(nameof(packageVersion));
+            _ = sources ?? throw new ArgumentNullException(nameof(sources));
+
+            bool atLeastOneSourceValid = false;
+            foreach (PackageSource source in sources)
+            {
+                _nugetLogger.LogDebug($"Searching {packageIdentifier}::{packageVersion} in {source.Source}.");
+                IEnumerable<IPackageSearchMetadata> foundPackages = await GetPackageMetadataAsync(source, packageIdentifier, includePrerelease:true, cancellationToken).ConfigureAwait(false);
```

Comment:
NuGet side we just parallelize these calls. There's no API but it should be easy to implement either like this https://github.com/NuGet/NuGet.Client/blob/f754a0da911a07cfac987df0fbdf5518692e3d5a/src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs#L68-L97 or by using dataflow.

Reply:
Thanks, will do that.